File: 3158/find-the-xor-of-numbers-which-appear-twice.py
# LeetCode 3158. Find the XOR of Numbers Which Appear Twice
# 找到 nums 裡「剛好出現2次」的數, 全部用 XOR 混起來。
# 直覺會利用「字典」, 記下「數字出現的次數」,看誰出現2次
class Solution:
    def duplicateNumbersXOR(self, nums: List[int]) -> int:
        table = {}  # 這裡用Python 的「字典」來做
        for num in nums:  # 全部數字都巡一次
            if num not in table:  # 若沒出現過
                table[num] = 1  # 就加到 table 裡
            else:  # 若曾出現過
                table[num] += 1  # 出現次數就+1
        ans = 0
        for num in table:  # 巡第2次, 看 table 裡的每個數
            if table[num]==2:  # 題目要的、剛好出現2次的數
                ans ^= num  # 照題目要求, 用 XOR 混起來
        return ans

File: 3158/find-the-xor-of-numbers-which-appear-twice.c
// LeetCode 3158. Find the XOR of Numbers Which Appear Twice
// 找到 nums 裡「剛好出現2次」的數, 全部用 XOR 混起來。
// 因數字很少(最多50個數), 且介於 1..50之間, 所以開個陣列, 這題就搞定了。
// 記下「數字出現的次數」,看誰出現2次
int duplicateNumbersXOR(int* nums, int numsSize) {
    int table[51] = {};  // 陣列從 0 開始, 所以要開 51格, 裝1..50之間的數
    for(int i=0; i<numsSize; i++) {
        table[nums[i]] ++;  // 數字出現次數 +1
    }
    int ans = 0;
    for(int i=1; i<=50; i++) {  // 巡第2次, 看 table 裡的每個數
        // 查看 table[i] 裡, 有誰「剛好出現2次」, 就要用 XOR 混到答案裡
        if(table[i]==2) ans = ans ^ i;
    }
    return ans;
}
